### PR TITLE
benchmarks: make benchmarks more stable

### PR DESCRIPTION
**Pull request by eeckstein**
The time per iteration should roughly stay constant and not depend on the number of iterations.

*) for benchmarks with a significant setup time, move the setup to the setUpFunction
*) some benchmarks wrongly executed the loop N+1 times ("0...N" instead of "0..<N")
*) added some blackHole calls to prevent the optimizer removing important parts of a benchmark
*) lowered the workload for the CSVParsing benchmarks because 1 single iteration took too long
*) moved the setup and teardown function calls outside the sample loop

For a few benchmarks this changed the benchmark score. I renamed those benchmarks so that they are now treated as "new" benchmarks.

### CODE REVIEW

**Comment by eeckstein** [comment]
@swift-ci smoke test

**Comment by eeckstein** [comment]
@swift-ci smoke benchmark

**Comment by eeckstein** [comment]
@palimondo I went ahead and fixed some of the benchmark issues.

**Comment by palimondo** [comment]
@eeckstein Good to see! Let me have a look at these today, but I need review the changes in more detail. 

On a cursory glance, I'd lower the workload in CVSParsing much lower still.
> **Note:** Getting the runtime under 2500 μs is very important for reducing noise. It gives you 1:4 chance the measurement will be interrupted by system scheduler (we can deal with that using the filtering I'm proposing). Any measurement of a single sample with runtime above 10ms (10 000 μs) is guaranteed to be interrupted by scheduler on Mac OS with a random delay error, that's accumulated into the sample.

But since that is the single modification to that benchmark, I'd delay changing it until we land the `BenchmarkDoctor` which will give us precise guidance. There will be a ton of these workload adjustments.

**Comment by eeckstein** [comment]
@swift-ci smoke test

**Comment by palimondo** [comment]
I think I can get to look at this in detail only over the weekend, but I had a slightly different pattern for taking the setup overhead out of the main loop in my mind…

I’m also a bit confused by the use of `blackHole` in some of the `setUp` functions here. What was the intention there?

**Comment by eeckstein** [comment]
@palimondo Makes sense, I changed the iteration time of the renamed benchmarks (like CSVParsing) to be < 2.5ms. In this PR I addressed all the benchmarks where I saw a problem for reducing the iteration count (and increasing the sample count). So this PR should give us a good starting point. 
(Whether to change a majority of other benchmarks is a separate discussion).

**Comment by eeckstein** [comment]
> I’m also a bit confused by the use of blackHole in some of the setUp functions here.

Global variables are initialized lazily. The blackHole in the setup functions make sure the initialization is done before the run-functions are called.

**Comment by palimondo** [comment]
I see. The pattern I was thinking to use was to have **optional** global, initialize in `setUp` but also set it to `nil` in `tearDown`, so that we don’t bloat memory as we progress through the Suite. Does that make sense?

File: benchmark/single-source/Array2D.swift
Comment (palimondo) [inline benchmark/single-source/Array2D.swift]: I wasn't sure what was the main point of this benchmark before, when I've noticed it has unstable memory consumption. From the [robust-microbench report](https://palimondo.github.io/robust-microbench/memory-use.html):
> The Array2D benchmark has significant memory use range of 7MB (3292 — 5101 pages or 12.9 MB — 19.9 MB)! It creates a 1K by 1K [[Int]], without reserving a capacity. The pure Int storage is at least 8 MB, plus there is some constant overhead per Array. I guess the variation depends on how big contiguous memory regions the allocator gets, while the arrays are growing when append is called and they sometimes need to be copied to new region. Though I’m not sure this is the point of the test and it maybe should be rewritten with reserved capacity for stable memory use.

Seeing you've taken the array initialization out into the `setUpFunction`… would you mind fixing it to use reserved capacity? How about something like this:
````swift
return Array(repeating: Array(0 ..< 1024), count: 1024)
````

File: benchmark/single-source/Array2D.swift
Comment (palimondo) [inline benchmark/single-source/Array2D.swift]: I think we should store these workload data structures in `Optional`s, so that we can release them in the `tearDownFunction`. Otherwise we're effectively leaking memory, as we go through the pre-commit benchmark suite. This single array here is 8MB.

**Comment by eeckstein** [comment]
That's a good point. But I think in most cases the data overhead is not that big and it does not really matter. Therefore I prefer the simpler method (of just referencing the global instead of an optional).

Actually I have an idea to simplify this setup/teardown procedure at all, which would solve this problem. I'll prepare a PR for this.

File: benchmark/single-source/ReversedCollections.swift
Comment (palimondo) [inline benchmark/single-source/ReversedCollections.swift]: I don't think renaming ReversedCollections is necessary. You are not changing the workload, just fixing a bug in test. (Unless it was meant as a technique to highlight them for review in the benchmark report...)

File: benchmark/single-source/ClassArrayGetter.swift
Comment (palimondo) [inline benchmark/single-source/ClassArrayGetter.swift]: Does bug fix warrant a name change?

**Review by palimondo — CHANGES_REQUESTED** [review]
Most of these changes LGTM, but I'd like to see the `Array2D` initialization with reserved capacity. 

I don't think its necessary to rename tests where the change in runtime is due to fixing bugs by adding `blackhole` in the right places. Before this, most of these tests were reporting just the timing of the [setup overhead](https://palimondo.github.io/robust-microbench/exclude-setup-overhead.html) anyway. This is really mostly a bugfix PR.

Can you do one more **smoke benchmark**, please?

File: benchmark/single-source/ArrayOfGenericPOD.swift
Comment (palimondo) [inline benchmark/single-source/ArrayOfGenericPOD.swift]: This one is 50% overhead according to my measurements. I guess some `blackhole`s are needed in there, too?

**Review by palimondo — CHANGES_REQUESTED** [review]

File: benchmark/single-source/ArrayOfGenericPOD.swift
Comment (eeckstein) [reply benchmark/single-source/ArrayOfGenericPOD.swift]: I didn't see any problems here, but it's a good idea to add blackHoles here.

File: benchmark/single-source/Array2D.swift
Comment (eeckstein) [reply benchmark/single-source/Array2D.swift]: I simplified it. But note that your elegant single-liner is not equivalent to the original construction - performance wise, because it reuses the same inner array. This then triggers copies-on-write in the main benchmark loop.

**Comment by eeckstein** [comment]
@palimondo Thanks for reviewing. I solved the problem of leaking memory through the global workloads with a new (and much simpler) mechanism for separating setup time from core benchmark time.

About renaming benchmarks: the purpose of renaming a benchmark is to not introduce false changes in the benchmark history. It has nothing to do if it's a bug fix or not. If the benchmark time changes (significantly) by changing the benchmark itself (and not the compiler) then it should be renamed.

**Comment by eeckstein** [comment]
@swift-ci smoke benchmark

**Review by palimondo — CHANGES_REQUESTED** [review]
I think this is solving a wrong problem. I'll expand in the followup comment.

Otherwise, the `blackhole` and range fixes, as well as the `Array2D` initialization look solid.

**Comment by palimondo** [comment]
@eeckstein, @gottesmm I think there is some misunderstanding of the setup overhead issue I've raised in [my report](https://palimondo.github.io/robust-microbench/exclude-setup-overhead.html), when I suggested it should to be removed:

> Some benchmarks need to perform additional setup work before their main workload. Historically, this was dealt with by sizing the main workload so that it dwarfs the setup, making it negligible. Setup is performed before the main work loop, and its impact is further lessened by amortizing it over the N measured iterations.

This, in and of itself, is completely unrelated to the stability of the measurement results. The suggestion to address setup overhead is predicated in the main proposed change: measuring all benchmarks with `--num-iters=1` and collecting a lot of samples (that will be filtered for noise). Since we'll no longer be measuring with `N>1`, the effect of setup becomes more pronounced, because it wouldn't be amortized over the high number of iterations (`N`) as is currently the case.

> The setup overhead is a systematic measurement error, that can be detected and corrected for, when measuring with different `num-iters`. Given two measurements performed with `i` and `j` iterations, that reported corresponding runtimes `ti` and `tj`, the setup overhead can be computed as follows:
> ````
> setup = (i * j * (ti - tj)) / (j - i)
> ````

In fact, to be most precise you should compute this with the lowest `N` possible: when `i=1` and `j=2`, this equation reduces to `setup = 2 * (t1 - t2)`. We can compute the ratio of setup to the main workload (`ratio = (setup / t1)`). I have included the [setup overhead detection in the `BenchmarkDoctor`](https://github.com/palimondo/swift/blob/capillary-dilation/benchmark/scripts/Benchmark_Driver#L263-L270), because it has proven itself to be an important metric of the benchmark health. Here's the beginning of a longer table from the report: 

| Benchmark | setup<br>overhead |
|---|---|
| ClassArrayGetter | [100%](chart.html?b=ClassArrayGetter&v=a10 "5974µs") |
| ReversedDictionary | [99%](chart.html?b=ReversedDictionary&v=a10 "48700µs") |
| ArrayOfGenericRef | [49%](chart.html?b=ArrayOfGenericRef&v=a10 "16292µs") | 
| ArrayOfPOD | [50%](chart.html?b=ArrayOfPOD&v=a10 "622µs") |
| ArrayOfGenericPOD | [50%](chart.html?b=ArrayOfGenericPOD&v=a10 "662µs") | 

...basically the task list for fixes in your 5th commit here (1b5ad3c). High setup overhead ratio is useful for detecting the case when compiler optimizations kicked-in and eliminated some part of the workload and judicious `blackhole` wrapping is necessary to correct it. 

----

Now, why and how should we exclude the setup overhead once we start to measure with `N=1`?
Keep in mind, we'll be trading `num-samples=1` with` num-iters=X`, for `num-samples=X` with `num-iters=1`, where `X` is the automatically computed scale factor to gather measurements for ~1s.

* Non-amortized setup overhead would increase reported runtime for that benchmark.
* We'd be paying setup overhead on each measurement, increasing overall time to run benchmarks.

So the whole point of moving the setup work that is detected as overhead into the `setUpFunction` is to maintain the status quo! Perform the setup work just once, before running the test, amortizing it over the whole measurement of a given benchmark. 

[Erik's initial attempt](https://github.com/apple/swift/pull/18096#issuecomment-406667786) here in an earlier, overwritten commit with `setUpFunction: { blackhole(workload) }` was moving in the right direction, except for the fact that lazily initialized globals were lingering in memory after the benchmark ran. This wasn't a problem in the "overhead-included" version as the workload was inside the scope of `runFunction` and went away when it finished running. 

I understand the motivation to make writing of benchmarks as easy as possible and that setting up an *optional global variable* that would be cleaned up in the `tearDownFunction` was looking like a more complicated solution… but the changes in the first 3 commits (35fc739, 24682ec, eb92982) that completely eliminate the `setUp` and `tearDown` functions are IMHO solving the wrong problem. The point of eliminating setup overhead isn't to pretend it doesn't exist (just excluding it from the measured time), but to maintain the status quo of amortizing setup over the whole measurement. We have [briefly touched on this issue](https://github.com/apple/swift/pull/12415#discussion_r201793128) in the review of PR #12415, where I've argued that the `setUp` and `tearDown` functions should be moved outside of the sampling loop. Does this explanation make my point more clear now?

Maybe we can go back to the original Eric's idea with the lazily initialized globals? What if the pattern was to make them IUOs (so that the main workflow body can continue to pretend its unchanged local variable) and do something like `tearDownFunction: { release(workload) } `?

----

But to reiterate, with the current measurement method, setup overhead plays absolutely no role. It becomes an issue, affecting about 10% of the benchmarks in the suite, only when we start measuring with `--num-iters=1`. I would caution against attempting to address it before we have [`BenchmarkDoctor`](https://github.com/palimondo/swift/blob/capillary-dilation/benchmark/scripts/Benchmark_Driver) in place to guide us.

**Comment by palimondo** [comment]
@eeckstein Could you please hold this until after we land #18124? There's looming merge conflict in `DriverUtils.swift`...

I understand the usability point of view, and agree that the ability to write benchmarks correctly is very important.  But the proposed change (removing `setUp` and `tearDown`) *is* functional, because the new `startBenchmarkTiming`, `endBenchmarkTiming` and `runBenchmark(_:workload:)` functions completely remove the ability to detect the setup overhead. They just hide it. This makes it impossible to validate the correctness of benchmarks in this regard. It doesn't isolate the overhead in one place, making it impossible to amortize, when we change the measurement method from *many-iterations* to  *many-samples*. Than we'll be paying the setup overhead too many times.

Ensuring that benchmarks are written correctly, adhering to all the assumptions of our testing process is the main point of introducing benchmark validation infrastructure that I'm proposing with [`BenchmarkDoctor`](https://github.com/palimondo/swift/blob/capillary-dilation/benchmark/scripts/Benchmark_Driver#L166). It would be possible to add checks for potential issues you've raised (validate the workload is destroyed and globals are used correctly).

**Comment by palimondo** [comment]
There are two equally valid approaches to benchmarking frameworks with regards to which part controls the time measurement: the **driver** or the **test**. In Swift Benchmarking Suite as it currently stands, it is the driver that controls the measurement and the benchmark supplies the function to be measured. In the other approach you hand the stopwatch to the test itself, which is exactly what the changes proposed in the first 3 commits here do. This would create a hybrid situation in SBS.

With the *test-controls-timing* approach it is possible to precisely control what get's measured. That is crucial if you want to measure effects with nanosecond precision. It's the approach taken by @lorentey's Attabench framework. But since everything is controlled from the test, the responsibility for correct use lies completely in the hands of test author.

With the *driver-controls-timing* approach we have more control over the measurement strategy as a whole process. The trade-off is that benchmarks have to give up fine-grained control and need to conform to the set of assumptions of this process. This is more like using unit testing framework, with special affordances for rare but sometimes necessary tasks like `setUp` and `tearDown`. When we add infrastructure to validate that benchmarks conform to the given set of assumptions, we'll have a **central point to monitor the quality** of the whole SBS. 

I believe that the second approach is better for the SBS, because it gives us leverage to measure, improve and control its quality. When the details of what and how get's measured are left underspecified, it's easy to deviate from the best practices and the result is set of benchmarks that take too long to run and report spurious changes.

**Comment by eeckstein** [comment]
@swift-ci smoke test

**Comment by palimondo** [comment]
But even if you worked around this issue by tracking the wall time inside SampleRunner in another global... giving the stopwatch to the hands of the test just hides the setup overheard. When we increate the sampling frequency, we'll be paying it all the time. We have to do the setup only once and amortize it across many measurements (whether we're talking about `num-iters` or `num-samples` doesn't matter). Currently the `setUpFunction` gives us a place to do that. 

If you think the ergonomics around its use in conjunction with `tearDownFunction` and not leaking the workloads are suboptimal, we can focus on that. But we giving stopwatch to the hands of the tests themselves will exacerbate the problem of correctly managing the setup overhead across the SBS.

**Comment by eeckstein** [comment]
> I don't see how this would work with the change as is

```
diff --git a/benchmark/utils/DriverUtils.swift b/benchmark/utils/DriverUtils.swift
index 5bed947501..c2e8bf04c9 100644
--- a/benchmark/utils/DriverUtils.swift
+++ b/benchmark/utils/DriverUtils.swift
@@ -277,2 +277,3 @@ class SampleRunner {
     endBenchmarkTiming()
+    let totalTime = _getTotalTime()
 
diff --git a/benchmark/utils/TestsUtils.swift b/benchmark/utils/TestsUtils.swift
index ee61a31e81..518e14accb 100644
--- a/benchmark/utils/TestsUtils.swift
+++ b/benchmark/utils/TestsUtils.swift
@@ -296,2 +296,3 @@ func getDiffTimeInNanoSeconds(start: TimeT, end: TimeT) -> UInt64 {
 var start_ticks: TimeT = TimeT()
+var start_run_function_ticks: TimeT = TimeT()
 var time_running = false
@@ -302,2 +303,3 @@ public func _resetBenchmarkTiming() {
   benchmarkNanoSeconds = 0
+  start_run_function_ticks = getTime()
 }
@@ -309,2 +311,6 @@ public func _getBenchmarkTiming() -> UInt64 {
 
+public func _getTotalTime() -> UInt64 {
+  return getDiffTimeInNanoSeconds(start: start_run_function_ticks, end: getTime())
+}
+
 /// Explicitly start benchmark timing within a run-function.
```

**Comment by palimondo** [comment]
Just to be sure I didn't overlook it before, the workaround to maintain the ability to detect overhead in the diff above, with `_getTotalTime()` function was not part of the commits in this PR yet, correct?

I agree the issue of setup overhead is very minor. I have been saying this in the forums ever since Michael got hung upon this point there. Which is why I don't think we have to rush in with a fix, without having the tools that measure it in the first place. It is severe for handful of test, very minor for vast majority of test that exhibit the overhead and it affects less than 10% of all benchmarks in the suite.

I think the current setup works just fine. The move to handing the stopwatch to the test in order to address setup overhead goes against the current design of the SBS and will make it harder to deal with it correctly, when the need arises.

**Comment by palimondo** [comment]
Thinking too much about the setup overhead is a case of premature optimization. I think if we stuck to the current design, where you write the benchmark in a single function and don't have to think about it until the overhead get's flagged as an issue by `BenchmarkDoctor` would be easiest. Then we just need a well documented pattern to deal with it, like the global optional with blackhole invoked in the `setUp` and a release in `tearDown`. 

I was [saying in the discussion](https://github.com/apple/swift/pull/12415#discussion_r201793128) of PR #12415, why it's safe to move them outside of the sampling look: the `setUpFunction` as currently deployed in the SBS (as you know, now that you've touched all of them in this PR) isn't used like the `setUp` and `tearDown` in unit tests, but like these methods on `TestCase`. That's because the driver is supplying the `num-iters` or`N`(as it is called when passed to the `runFunction`). If a test needs a fresh instance of workload data, it needs to get it for every *iteration*. So if it builds some expensive data structure in the `setUp`, it still needs to copy it into another variable at the beginning of its inner loop(N).

All I was trying to say is that we should maintain the support for `setUp` in the benchmarking framework API and not hand the ability to hide setup overhead to every benchmark author. Of course, it is possible to work around such design by instrumenting driver to measure things twice. But I don't think it's the right approach. Semantically correct place for putting a test setup is the `setUp` method.

The pattern with lazy global variable you came up with was great. The potential memory leak we were discussing also isn't that big of a deal. When running test via `Benchmark_Driver`, they are all executed individually (independent `Benchmark_O` invocations), so it doesn't matter there. I don't know how you are using the `Benchmark_O` internally, tough. If you are running in a mode where it runs through the whole SBS pre-commit suite (no filters or tags), or with a `--tags` where it executes a bunch of tests in one run, it might be slowly accumulating all the lazily initialized globals, if you didn't clean it up correctly. But that also doesn't seem like such a big problem, only a minor pedantic point.

To avoid that minor issue, we can use `Optional`s and release them in teardown. The ergonomics of this pattern can be polished further to enforce the use of implicitly unwrapped optionals if correct application of the pattern becomes a problem.

**Comment by eeckstein** [comment]
ok, I see you have a strong opinion on this. I removed the commits which introduce the new setup mechanism (and updated the other commits). With the old method I agree with you that the setup/teardown function calls should be moved out of the sample loop. I added a commit for this.

**Comment by eeckstein** [comment]
@swift-ci smoke benchmark

File: benchmark/single-source/ArrayOfGenericPOD.swift
Comment (palimondo) [inline benchmark/single-source/ArrayOfGenericPOD.swift]: I'm confused about all these benchmarks that have the `// should be a nop` comments. What are they supposed to measure?

Comment (eeckstein) [reply benchmark/single-source/ArrayOfGenericPOD.swift]: I _think_ it means that it's expected that the optimizer removes everything. I'm not sure if this makes sense.

Comment (eeckstein) [reply benchmark/single-source/ArrayOfGenericPOD.swift]: In general, it's better to replace such benchmarks (which only test if the optimizer removes everything) by a lit test.

**Comment by eeckstein** [comment]
@swift-ci smoke benchmark

File: benchmark/single-source/StringComparison.swift.gyb
Comment (palimondo) [inline benchmark/single-source/StringComparison.swift.gyb]: This whole file is absolutely beautiful. I haven't seen this benchmark before, as I was stuck on a November 2017 fork. This is exactly the correct pattern to emulate when you have a significant setup overhead. Excellent engineering all over here!

**Review by palimondo — APPROVED** [review]
This LGTM! Thank you for hearing me out!

**Comment by eeckstein** [comment]
@swift-ci smoke test and merge

**Comment by eeckstein** [comment]
@swift-ci smoke test and merge

**Comment by eeckstein** [comment]
@swift-ci smoke test OS X platform

**Comment by eeckstein** [comment]
@swift-ci smoke test macOS

**Pull request merged**
